toolchain/language_server/context.cpp

```diff
@@ -6,45 +6,143 @@
 
 #include <memory>
 
+#include "common/check.h"
+#include "common/raw_string_ostream.h"
 #include "toolchain/base/shared_value_stores.h"
-#include "toolchain/diagnostics/null_diagnostics.h"
+#include "toolchain/check/check.h"
+#include "toolchain/diagnostics/diagnostic.h"
+#include "toolchain/diagnostics/diagnostic_consumer.h"
 #include "toolchain/lex/lex.h"
 #include "toolchain/lex/tokenized_buffer.h"
 #include "toolchain/parse/parse.h"
 #include "toolchain/parse/tree_and_subtrees.h"
 
 namespace Carbon::LanguageServer {
 
-auto Context::File::SetText(Context& context, llvm::StringRef text) -> void {
+class PublishDiagnosticConsumer : public DiagnosticConsumer {
+ public:
+  explicit PublishDiagnosticConsumer(Context* context,
+                                     const clang::clangd::URIForFile& uri,
+                                     std::optional<int64_t> version)
+      : context_(context), params_{.uri = uri, .version = version} {}
+
+  auto params() -> llvm::json::Value { return params_; }
+
+  auto HandleDiagnostic(Diagnostic diagnostic) -> void override {
+    const auto& message = diagnostic.messages[0];
+    if (message.loc.filename != params_.uri.file()) {
+      // `pushDiagnostic` requires diagnostics to be associated with a location
+      // in the current file. Suppress diagnostics rooted in other files.
+      // TODO: Consider if there's a better way to handle this.
+      RawStringOstream stream;
+      StreamDiagnosticConsumer consumer(&stream);
+      consumer.HandleDiagnostic(diagnostic);
+
+      CARBON_DIAGNOSTIC(LanguageServerDiagnosticInWrongFile, Warning,
+                        "dropping diagnostic in {0}:\n{1}", std::string,
+                        std::string);
+      context_->file_emitter().Emit(
+          params_.uri.file(), LanguageServerDiagnosticInWrongFile,
+          message.loc.filename.str(), stream.TakeStr());
```

Comment:
Is this recursing back into this same function to produce the error, or is this diagnostic going somewhere else? If this recurses back here, would it make sense to directly push a diagnostic instead?